esphome/components/waveshare_epaper/weact_2p9in3c.cpp

```diff
@@ -0,0 +1,169 @@
+static const uint8_t RAM_Y_RANGE[] = {0x45, 0x00, 0x00, (uint8_t) HEIGHT - 1, (uint8_t) (HEIGHT >> 8)};
+static const uint8_t RAM_X_POS[] = {0x4E, 0x00};  // Always start at 0
+static const uint8_t RAM_Y_POS = 0x4F;
+
+#define SEND(x) this->cmd_data(x, sizeof(x))
+
+// Basics
+
+int WeActEPaper2P9In3C::get_width_internal() { return WIDTH; }
+int WeActEPaper2P9In3C::get_height_internal() { return HEIGHT; }
+uint32_t WeActEPaper2P9In3C::idle_timeout_() { return 2500; }
+
+void WeActEPaper2P9In3C::dump_config() {
+  LOG_DISPLAY("", "WeAct E-Paper (3 Color)", this)
+  ESP_LOGCONFIG(TAG, "  Model: 2.90in Red+Black");
+  LOG_PIN("  CS Pin: ", this->cs_)
+  LOG_PIN("  Reset Pin: ", this->reset_pin_)
+  LOG_PIN("  DC Pin: ", this->dc_pin_)
+  LOG_PIN("  Busy Pin: ", this->busy_pin_)
+  LOG_UPDATE_INTERVAL(this)
+}
+
+// Device lifecycle
+
+void WeActEPaper2P9In3C::setup() {
+  setup_pins_();
+  delay(20);
+  this->send_reset_();
+  // as a one-off delay this is not worth working around.
+  delay(100);  // NOLINT
+  this->wait_until_idle_();
+  this->command(SW_RESET);
+  this->wait_until_idle_();
+
+  SEND(DRV_OUT_CTL);
+  SEND(DATA_ENTRY);
+  SEND(BORDER_FULL);
+  SEND(TEMP_SENS);
+  SEND(DISPLAY_UPDATE);
+
+  this->wait_until_idle_();
+}
+
+void WeActEPaper2P9In3C::send_reset_() {
+  if (this->reset_pin_ != nullptr) {
+    this->reset_pin_->digital_write(false);
+    delay(2);
+    this->reset_pin_->digital_write(true);
+  }
+}
+
+// must implement, but we override setup to have more control
+void WeActEPaper2P9In3C::initialize() {}
+
+void WeActEPaper2P9In3C::deep_sleep() { SEND(SLEEP); }
+
+// Pixel stuff
+
+// t and b are y positions, i.e. line numbers.
+void WeActEPaper2P9In3C::set_window_(int t, int b) {
+  SEND(RAM_X_RANGE);
+  SEND(RAM_Y_RANGE);
+  SEND(RAM_X_POS);
+
+  uint8_t buffer[3];
+  buffer[0] = RAM_Y_POS;
+  buffer[1] = (uint8_t) t % 256;
+  buffer[2] = (uint8_t) (t / 256);
+  SEND(buffer);
+}
+
+// send the buffer starting on line `top`, up to line `bottom`.
+void WeActEPaper2P9In3C::write_buffer_(int top, int bottom) {
+  auto width_bytes = this->get_width_internal() / 8;
+  auto offset = top * width_bytes;
+  auto length = (bottom - top) * width_bytes;
+
+  this->wait_until_idle_();
+  this->set_window_(top, bottom);
+
+  this->command(WRITE_BLACK);
+  this->start_data_();
+  this->write_array(this->buffer_ + offset, length);
+  this->end_data_();
+
+  offset += this->get_buffer_length_() / 2;
+  this->command(WRITE_COLOR);
+  this->start_data_();
+  this->write_array(this->buffer_ + offset, length);
+  this->end_data_();
+}
+
+void HOT WeActEPaper2P9In3C::draw_absolute_pixel_internal(int x, int y, Color color) {
+  if (x >= this->get_width_internal() || y >= this->get_height_internal() || x < 0 || y < 0)
+    return;
+
+  const uint32_t buf_half_len = this->get_buffer_length_() / 2u;
+
+  const uint32_t pos = (x + y * this->get_width_internal()) / 8u;
+  const uint8_t subpos = x & 0x07;
```

Comment:
Should pre-calculate `(0x80 >> subpos) ` here:
```suggestion
  const uint8_t sub_bit = 0x80 >> (x & 0x07);
```
And replace in subsequent code accordingly.